src/HttpServer.php

```diff
@@ -103,42 +103,36 @@ public function handleConnection(ConnectionInterface $connection)
     /** @internal */
     public function handleBody(RequestInterface $request, ConnectionInterface $connection)
     {
-        $bodyBuffer = '';
-        $that = $this;
-
-        $protection = new CloseProtectionStream($connection);
-
         if ($this->isChunkedEncodingActive($request)) {
-            $chunkedDecoder = new ChunkedDecoder($protection);
-            $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $connection, $that) {
-                $bodyBuffer .= $chunk;
-                if (strlen($chunk) == 0) {
-                    $request = $request->withBody(RingCentral\Psr7\stream_for($bodyBuffer));
-                    $that->handleRequest($connection, $request);
-                }
-            });
+            // Add ChunkedDecoder to stream
+            $chunkedDecoder = new ChunkedDecoder($connection);
+            $bodyStream = new HttpBodyStream($chunkedDecoder);
+            $request = $request->withBody($bodyStream);
+            $this->handleRequest($connection, $request);
             return;
         }
 
-        $contentLengthArray = $request->getHeader('Content-Length');
-
-        if (!$request->hasHeader('Content-Length') || $contentLengthArray[0] == 0) {
+        if (!$request->hasHeader('Content-Length')) {
+            // Request hasn't defined 'Content-Length' will ignore rest of the request
+            // and ends the stream
+            $bodyStream = new HttpBodyStream($connection);
+            $request = $request->withBody($bodyStream);
             $this->handleRequest($connection, $request);
+            $bodyStream->emit('end', array());
```

Comment:
Not sure if this is tested, but could the connection keep emitting `data` events here?